src/LazyArray.h

```diff
@@ -0,0 +1,127 @@
+
+#pragma once
+
+/** A dynamic array that defers allocation to the first modifying access.
+Const references from the array before allocation will all be to the same default constructed value.
+It is therefore important that default constructed values are indistinguishable from each other. */
+template <typename T>
+class cLazyArray
+{
+	static_assert(!(std::is_reference<T>::value || std::is_array<T>::value),
+		"cLazyArray<T>: T must be a value type");
+	static_assert(std::is_default_constructible<T>::value,
+		"cLazyArray<T>: T must be default constructible");
+public:
+	using value_type = T;
+	using pointer = T *;
+	using const_pointer = const T *;
+	using reference = T &;
+	using const_reference = const T &;
+	using size_type = int;
+	using iterator = pointer;
+	using const_iterator = const_pointer;
```

Comment:
Using the pointers works for iterators, but I don't like the side effect of const iterators initializing the data structure.
Maybe this explains, why you only use the for each loop only sometimes.

So I would suggest a seperate class for (at least) the `const_iterator`. This class stores the current position and a pointer to `cLazyArray` and calls the `GetAt`-Method to access the elements.

This should allow the consistent usage of the for each loop. Furthermore it would remove the `IsStorageAllocated` call from most (not all!) access only functions, like `CopyToItemStack`.

Reply:
My main point is the consistency. I don't want to see other people refactoring the `for(int i)` loop to an for each loop and causing implicit memory allocation (thus break the goals of refactoring). And commenting every `for(int i)` is just bad.

So either implement the [iterator](http://en.cppreference.com/w/cpp/concept/Iterator) (I don't need a random access iterator, just something with an `operator++` and an `operator*`). I would also do the work of implementing it, if you don't want to implement it 🎢 .

Or if you don't want to go too far in this PR, just drop the whole iterator concept out of `cLazyArray`! Then it can be later implemented in a seperate PR. Furthermore this would also give the benefits of no need to change `CopyToItemStack` and reduce the usage of `IsStorageAllocated`.

Reply:
Thanks for clarification! Just looked at all the places and the diff has just hidden alle the usages of the index.